### PR TITLE
Fix KeyInfoTest.ImportKeyNode

### PR DESCRIPTION
**Pull request by krwq**
Fixes https://github.com/dotnet/corefx/issues/16779

Previous version of the test was not Debuggable so I've split it so that we test individual components separately. Also added additional test for DSAKeyValue

### CODE REVIEW

File: src/System.Security.Cryptography.Xml/tests/KeyInfoTest.cs [hunk 1]
```diff
@@ -195,18 +196,139 @@ public void Complex()
...
+                    byte[] counter = Convert.FromBase64String(dsaPgenCounter);
+                    Assert.InRange(counter.Length, 1, 4);
+                    int counterVal = 0;
+                    for (int j = 0; j < counter.Length; j++)
+                    {
+                        counterVal <<= 8;
+                        counterVal |= counter[j];
+                    }
+
+                    Assert.Equal(counterVal, dsaParams.Counter);
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Xml/tests/KeyInfoTest.cs]: I expect this to fail on Linux and/or macOS.  On Linux we don't roundtrip Counter.  Counter and Seed should have the same nullarity, though.  (Counter == null => Seed == null; Seed == null => Counter == null).

Similar to J, if it's present it should be right.

File: src/System.Security.Cryptography.Xml/tests/KeyInfoTest.cs [hunk 2]
```diff
@@ -195,18 +196,139 @@ public void Complex()
...
+                    Assert.Equal(Convert.FromBase64String(x509cert), cert.GetRawCertData());
+                }
+                else
+                {
+                    Assert.True(false, $"Unexpected clause type: {clause.GetType().FullName}");
+                }
+            }
+
+            // 0x1f = b11111, number of ones = 5
+            Assert.Equal(pathsCovered, 0x1f);
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Xml/tests/KeyInfoTest.cs]: Does this really need to be one mondo test, or can/should it be broken up into individual tests?

Comment (krwq) [inline src/System.Security.Cryptography.Xml/tests/KeyInfoTest.cs]: This test is testing multiple clauses, we already have individual paths tested

File: src/System.Security.Cryptography.Xml/tests/KeyInfoTest.cs [hunk 3]
```diff
@@ -274,18 +274,28 @@ public void ImportKeyNode()
...
+                            counterVal <<= 8;
+                            counterVal |= counter[j];
+                        }
+
+                        Assert.Equal(counterVal, dsaParams.Counter);
+                    }
+                    else
+                    {
+                        Assert.Null(dsaParams.Seed);
+                        Assert.Equal(default(int), dsaParams.Counter);
```
Comment (krwq) [inline src/System.Security.Cryptography.Xml/tests/KeyInfoTest.cs]: hope this is right value, there is no null here unfortunatelly

**Comment by krwq** [comment]
@bartonjs seems like some OSX specific failure (new test case):
```
        System.Security.Cryptography.CryptographicException : An error occurred parsing the key components
        ---- System.Security.Cryptography.CryptographicException : Specified key is not a valid size for this algorithm.
        Stack Trace:
              at System.Security.Cryptography.Xml.DSAKeyValue.LoadXml(XmlElement value)
              at System.Security.Cryptography.Xml.KeyInfo.LoadXml(XmlElement value)
              at System.Security.Cryptography.Xml.Tests.DSAKeyValueTest.ImportDSAKeyValue()
           ----- Inner Stack Trace -----
              at System.Security.Cryptography.DSA.DSAImplementation.DSASecurityTransforms.ImportParameters(DSAParameters parameters)
              at System.Security.Cryptography.Xml.DSAKeyValue.LoadXml(XmlElement value)
```

Is this an expected failure?

**Comment by bartonjs** [comment]
> Is this an expected failure?

It looks like you have a DSA key larger than 1024-bit.  Apple only supports FIPS 186-2 DSA, which is limited to 1024-bit.

Should be happier with a smaller key :smile;

File: src/System.Security.Cryptography.Xml/tests/DSAKeyValueTest.cs [hunk 4]
```diff
@@ -123,5 +123,51 @@ public void LoadXml_Null()
...
+
+        [Fact]
+        public void ImportDSAKeyValue()
+        {
+            string p = "kkos0S7rZLGluCzOcQ/S9L79be47MxRp3kB07vkX+iin4daLcSo1pg/K9aP0UJzZAKUAyT8uj7jci6c5sfVWdqmlRqtm0mwnfh88JgmH/Eu5LkMrUAV2gDQ8Cs8ihd2UbbQ0s47P65qrINVpkHaNOFJbz8vcSP598kNLg73QKh0=";
+            string q = "l4A4DwlgltuK00dcNNW5Yw2SbcM=";
+            string g = "NiI1mc0jI7JQa21Pbxo2e87cJtUfO1kCor0LHbOk1T0W5h6uhIgdNwlmE/ARTqfcw+RamtHTEI9HfxTg1TCNPQnWYenEIOGaot0qRZs2a2WzD2xjictcEA2D8up0sqpK/qTcnNV9g6hcc3HAm9mU6DjQ21MOsIoYbQB1VW91FrA=";
+            string y = "Q2vVSH6O8/v6jJq4r3PQaCLKThnbRXxluEq9mEY0XnXTwefwTJ67fUcCxhJJhwh157gQN2LWmte5gnM11VJMWT07DLIGPb7he2wPRGmgnS4nkChGgJGKXss5unzFkX4m+BxCCXZrClrJyahlZWAJY5CbLanzHSZQtRHIDdINxf4=";
+            string seed = "zfYSHDy431BKUVSpA6KqGnCsbrE=";
+            string pgenCounter = "sg==";
```
Comment (tintoy) [inline src/System.Security.Cryptography.Xml/tests/DSAKeyValueTest.cs]: From memory, didn't we run into an edge case where pgenCounter was more than 1 byte long? Might be worth testing... :)

Comment (krwq) [inline src/System.Security.Cryptography.Xml/tests/DSAKeyValueTest.cs]: @tintoy do you have some more info? We do not support DSA generation on OSX and importing only works with keys up to 1024 bits. The only time I was seeing PgenCounter to have more than one byte is when I generated it on Desktop and key size was set to some larger value - I haven't generated too many keys though

Comment (tintoy) [inline src/System.Security.Cryptography.Xml/tests/DSAKeyValueTest.cs]: I'll try to dig it up; I remember the original logic for decoding pgenCounter was not symmetric with the logic for encoding it (it simply used the first byte and ignored the rest).

Comment (tintoy) [inline src/System.Security.Cryptography.Xml/tests/DSAKeyValueTest.cs]: Ah - found it! 19a367c8ca69871248f5cf08f34f257bdcc2ab23

Comment (krwq) [inline src/System.Security.Cryptography.Xml/tests/DSAKeyValueTest.cs]: Ok @tintoy, I'll check if I can generate some better key

**Comment by krwq** [comment]
@tintoy this should cover the corner case

**Pull request merged**